File: mac/Tests/DMShotTests/UpdatePromptTests.swift
import XCTest
@testable import DMShot

/// The gate in front of the active update prompt (spec 2026-08-02).
final class UpdatePromptTests: XCTestCase {
    private let now = Date(timeIntervalSince1970: 1_000_000)
    private func ready(_ v: String = "1.2.3") -> UpdateState { .readyToInstall(version: v) }

    func testOnlyPromptsWhenAnUpdateIsDownloaded() {
        for state: UpdateState in [.idle, .checking, .upToDate, .disabled,
                                   .available(version: "1.2.3", notes: []),
                                   .downloading(percent: 40), .extracting,
                                   .error(message: "nope")] {
            XCTAssertEqual(UpdatePrompt.decide(state: state, now: now, snooze: nil, busy: false), .none)
        }
    }

    func testPromptsWhenReadyAndFree() {
        XCTAssertEqual(UpdatePrompt.decide(state: ready(), now: now, snooze: nil, busy: false),
                       .show(version: "1.2.3"))
    }

    func testBusyDefersRatherThanDrops() {
        XCTAssertEqual(UpdatePrompt.decide(state: ready(), now: now, snooze: nil, busy: true), .wait)
    }

    func testActiveSnoozeStaysQuiet() {
        let snooze = UpdateSnooze(version: "1.2.3", until: now.addingTimeInterval(60))
        XCTAssertEqual(UpdatePrompt.decide(state: ready(), now: now, snooze: snooze, busy: false), .none)
    }

    func testExpiredSnoozePromptsAgain() {
        let snooze = UpdateSnooze(version: "1.2.3", until: now.addingTimeInterval(-1))
        XCTAssertEqual(UpdatePrompt.decide(state: ready(), now: now, snooze: snooze, busy: false),
                       .show(version: "1.2.3"))
    }

    func testNewerVersionBreaksTheSnooze() {
        let snooze = UpdateSnooze(version: "1.2.3", until: now.addingTimeInterval(86_400))
        XCTAssertEqual(UpdatePrompt.decide(state: ready("1.2.4"), now: now, snooze: snooze, busy: false),
                       .show(version: "1.2.4"))
    }

    func testSnoozeIsCheckedBeforeBusy() {
        // A snoozed update must not park the evaluation timer in `.wait` forever.
        let snooze = UpdateSnooze(version: "1.2.3", until: now.addingTimeInterval(60))
        XCTAssertEqual(UpdatePrompt.decide(state: ready(), now: now, snooze: snooze, busy: true), .none)
    }

    func testConstants() {
        XCTAssertEqual(UpdatePrompt.snoozeDuration, 24 * 60 * 60)
        XCTAssertEqual(UpdatePrompt.evaluationInterval, 60)
    }
}

File: windows/DMShot.Tests/UpdatePromptTests.cs
using DMShot.Update;
using Xunit;

namespace DMShot.Tests;

/// The gate in front of the active update prompt (spec 2026-08-02).
/// Mirrors mac/Tests/DMShotTests/UpdatePromptTests.swift case for case.
public class UpdatePromptTests
{
    private static readonly DateTimeOffset Now = new(2026, 8, 2, 12, 0, 0, TimeSpan.Zero);
    private static UpdateState Ready(string v = "1.2.3") => UpdateState.ForReadyToInstall(v);

    [Fact]
    public void OnlyPromptsWhenAnUpdateIsDownloaded()
    {
        UpdateState[] others =
        {
            UpdateState.Idle, UpdateState.Checking, UpdateState.UpToDate, UpdateState.Disabled,
            UpdateState.ForAvailable("1.2.3", Array.Empty<ChangelogVersion>()),
            UpdateState.ForDownloading(40), UpdateState.ForError("nope"),
        };
        foreach (var state in others)
            Assert.Equal(UpdatePromptAction.None, UpdatePrompt.Decide(state, Now, null, busy: false).Action);
    }

    [Fact]
    public void PromptsWhenReadyAndFree()
    {
        var d = UpdatePrompt.Decide(Ready(), Now, null, busy: false);
        Assert.Equal(UpdatePromptAction.Show, d.Action);
        Assert.Equal("1.2.3", d.Version);
    }

    [Fact]
    public void BusyDefersRatherThanDrops()
        => Assert.Equal(UpdatePromptAction.Wait, UpdatePrompt.Decide(Ready(), Now, null, busy: true).Action);

    [Fact]
    public void ActiveSnoozeStaysQuiet()
    {
        var snooze = new UpdateSnooze("1.2.3", Now.AddMinutes(1));
        Assert.Equal(UpdatePromptAction.None, UpdatePrompt.Decide(Ready(), Now, snooze, busy: false).Action);
    }

    [Fact]
    public void ExpiredSnoozePromptsAgain()
    {
        var snooze = new UpdateSnooze("1.2.3", Now.AddSeconds(-1));
        Assert.Equal(UpdatePromptAction.Show, UpdatePrompt.Decide(Ready(), Now, snooze, busy: false).Action);
    }

    [Fact]
    public void NewerVersionBreaksTheSnooze()
    {
        var snooze = new UpdateSnooze("1.2.3", Now.AddHours(24));
        var d = UpdatePrompt.Decide(Ready("1.2.4"), Now, snooze, busy: false);
        Assert.Equal(UpdatePromptAction.Show, d.Action);
        Assert.Equal("1.2.4", d.Version);
    }

    [Fact]
    public void SnoozeIsCheckedBeforeBusy()
    {
        // A snoozed update must not park the evaluation timer in Wait forever.
        var snooze = new UpdateSnooze("1.2.3", Now.AddMinutes(1));
        Assert.Equal(UpdatePromptAction.None, UpdatePrompt.Decide(Ready(), Now, snooze, busy: true).Action);
    }

    [Fact]
    public void SnoozeFromIgnoresHalfWrittenSettings()
    {
        Assert.Null(UpdatePrompt.SnoozeFrom("", Now));
        Assert.Null(UpdatePrompt.SnoozeFrom("1.2.3", null));
        Assert.NotNull(UpdatePrompt.SnoozeFrom("1.2.3", Now));
    }

    [Fact]
    public void MatchesTheMacConstants()
    {
        Assert.Equal(TimeSpan.FromHours(24), UpdatePrompt.SnoozeDuration);
        Assert.Equal(TimeSpan.FromSeconds(60), UpdatePrompt.EvaluationInterval);
    }
}
